Defining changed (issue #1167) (#1311)
ansible/ansible-documentation: docs/docsite/rst/playbook_guide/playbooks_handlers.rst

Sometimes you want a task to run only when a change is made on a machine. For example, you may want to restart a service if a task updates the configuration of that service, but not if the configuration is unchanged. Ansible uses handlers to address this use case. Handlers are tasks that only run when notified.

.. note::

>     Ansible sets "changed" status by itself, but you can also define the rules using ``changed_when``. See :ref:`override_the_changed_result` to determine it explicitly.

Reviewer: Rather than a note it might be worth adding a section or extending an existing section, like "Controlling when handlers run". As it is now, this doesn't really explain how `changed_when` relates to handlers.

This could be useful in a situation where you want to control when a handler runs by manually setting the "changed" status. I guess there could be instances when you'd want to notify a handler if the main task doesn't make any changes too.

The `playbooks_handlers.rst` file has some examples of notifying handlers to restart services so you could maybe keep it consistent with those and add something along the lines of:

```
- name: Copy httpd configuration
   ansible.builtin.copy:
     src: ./new_httpd.conf
     dest: /etc/httpd/conf/httpd.conf
   # The task is always reported as changed
   changed_when: True
   notify: Restart apache
```

This could be useful to ensure the handler restarts the service every time the new configuration gets copied. Maybe if the src and dest files are identical it wouldn't be marked as a change. You could probably confirm that with a quick test.

There is only one global scope for handlers (handler names and listen topics) regardless of where the handlers are defined. This also includes handlers defined in roles.

Controlling the notifing of handlers by defining "changed"

> ----------------------------------------------------------

Reviewer: This heading reads a little wordy to me. I don't think it would be good to make this section too repetitive with the `Defining “changed”` section. Maybe a heading like this:

"Defining when tasks change"

> By using ``changed_when`` keyword, you define when a task has "changed" status. This means you directly determine when ``notify`` is triggered so it executes the handlers.

Reviewer: I would combine these two sentences and maybe shorten them a little. Consider:

You can control when handlers are notified about task changes using the `changed_when` keyword.

This puts the focus on the more relevant details to this part of the docs, namely controlling when handlers get notified about changes. In the original first sentence it reads more like documentation about the `changed_when` keyword itself.

> In following example it is ensured that the handler restarts the service every time the new configuration gets copied:

Reviewer: ```suggestion
In the following example, the handler restarts the service each time the configuration file is copied:
```

.. code-block:: yaml

    - name: Copy httpd configuration
        ansible.builtin.copy:
          src: ./new_httpd.conf
          dest: /etc/httpd/conf/httpd.conf
        # The task is always reported as changed
        changed_when: True
        notify: Restart apache

> 

Reviewer: Please add `tasks:` to this example and fix the indentation.

Also you can execute handler for example based on command return value:

.. code-block:: yaml

    tasks:
      - name: Task with handled command result
        ansible.builtin.command: /bin/utils/check_command
        register: result
        changed_when: result.rc == 2 # something is not properly initialized
        notify: Initialize files

    handlers:
      - name: Initialize files

>         ...

Reviewer: This next example is OK but I think it is a little superfluous here. The example isn't consistent with the other examples, in that it isn't really about a service. This fits better with the section that you link to at the end so I'd just leave it out and keep it brief.
